cirq-google/cirq_google/engine/stream_manager.py

```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
+            try:
+                current_request.message_id = self._generate_message_id()
+                response_future = self._response_demux.subscribe(current_request.message_id)
+                await self._request_queue.put(current_request)
+                response = await response_future
+
+            # Broken stream
+            except google_exceptions.GoogleAPICallError as e:
+                if _is_retryable_error(e):
+                    # Retry
+                    self._response_demux.unsubscribe(current_request.message_id)
+                    current_request = get_result_request
+                    continue
+                    # TODO(#5996) add exponential backoff
+                raise e
+
+            # Either when this request is canceled or the _manage_stream() loop is canceled.
+            except asyncio.CancelledError:
+                if response_future is not None:
+                    response_future.cancel()
+                    await self._cancel(job.name)
+                raise
+
+            # Response handling
+            if 'result' in response:
+                return response.result
+            elif 'job' in response:
+                return response.job
+            elif 'error' in response:
+                current_request = _decide_retry_request_or_raise(
+                    response.error,
+                    current_request,
+                    create_program_and_job_request,
+                    create_job_request,
+                )
+                continue
+            else:
+                raise ValueError(
+                    'The Quantum Engine response type is not recognized by this client. '
+                    'This may be due to an outdated version of cirq-google'
+                )
+
+    async def _cancel(self, job_name: str) -> None:
+        await self._grpc_client.cancel_quantum_job(quantum.CancelQuantumJobRequest(name=job_name))
+
+    def _generate_message_id(self) -> str:
+        message_id = str(self._next_available_message_id)
+        self._next_available_message_id += 1
+        return message_id
+
+
+def _decide_retry_request_or_raise(
+    error: quantum.StreamError,
+    current_request,
+    create_program_and_job_request,
+    create_job_request: quantum.QuantumRunStreamRequest,
+):
+    """Decide whether the given stream error is retryable.
+
+    If it is, returns the next stream request to send upon retry. Otherwise, raises an error.
+    """
+    if error.code == Code.PROGRAM_DOES_NOT_EXIST:
+        if 'create_quantum_job' in current_request:
+            return create_program_and_job_request
+    elif error.code == Code.PROGRAM_ALREADY_EXISTS:
+        if 'create_quantum_program_and_job' in current_request:
+            raise ProgramAlreadyExistsError(
+                current_request.create_quantum_program_and_job.quantum_program.name
+            )
+    elif error.code == Code.JOB_DOES_NOT_EXIST:
+        if 'get_quantum_result' in current_request:
```

Comment:
Can you talk through this error mechanism? The best cause I can imagine is that the job was deleted, but then we shouldn't create a new one.

If that's true, then we wouldn't call `create_job_request` anywhere, which also makes the`PROGRAM_DOES_NOT_EXIST` error handling obsolete.

Reply:
Good point. What about the case where `CreateProgramAndJobRequest` isn't sent properly before the stream breaks? A `GetResultRequest` will be issued but the job doesn't exist. Unfortunately this is indistinguishable from the case where the job was deleted. I would say it's OK for the job to be recreated because if a user truly wants to stop a job execution, they could cancel the future returned by `submit()`. WDYT?